README.md
<!--
 * @Author: wangyunbo
 * @Date: 2021-05-19 23:57:46
 * @LastEditors: wangyunbo
 * @LastEditTime: 2021-05-21 17:33:14
 * @Description: file content
 * @FilePath: \css-variables-hints\README.md
-->

# css-variable-hints

Simple extension for geting hints for css-variables from selected files. You can trigger it by typing `--`.
一个简单的`css-variables`变量的提示插件,通过输入`--`触发。

![Demo](https://github.com/airbender92/css-variables-hints/blob/master/img/demo.gif)

Be sure that you have one of supported language modes selected in right bottom of your vscode window:
* css
* postcss
* scss
* less
* vue


## Extension Configuration
Create or modify file `.vscode/settings.json`. Folder `.vscode` should be in your workspace root directory.

在项目的`.vscode/settings.json`文件中配置css变量文件的路径数组，如下所示：
Minimal configuration file:
```
{
  ...
  // 本地资源变量文件路径列表
  "cssVarriablesHints.files": ["src\\theme\\variables.scss"],
  // 远端资源变量文件配置列表
  "cssVarriablesHints.remoteFiles": [{
    "projectId": 8888  //gitlab项目id号,
    "filePath": "lib/export.scss", // gitlab上变量资源的文件路径
    "token": "access token", // token
    "origin": "http://xxxx:8888" // origin
  } ],
  ...
};
```
`cssVarriablesHints.files` and `cssVarriablesHints.remoteFiles` are not required at the same time.
This extension contributes the following configuration parameters:

* `files`: array of paths to files with css variables    

![Demo](img/demo.gif) 
### 用法：   

1. 搜索插件名：`css-variables-hints`, 并安装。    
2. 修改项目根目录下的`.vscode/settings.json`：        
```javascript
{
  ...
  // 本地资源变量文件路径列表
  "cssVarriablesHints.files": ["src\\theme\\variables.scss"],
  // 远端资源变量文件配置列表
  "cssVarriablesHints.remoteFiles": [{
    "projectId": 8888  //gitlab项目id号,
    "filePath": "lib/export.scss", // gitlab上变量资源的文件路径
    "token": "access token", // token
    "origin": "http://xxxx:8888" // origin
  } ],
  ...
};
```
`cssVarriablesHints.files` 和 `cssVarriablesHints.remoteFiles` 可都配置也可任选一个配置.    
3. 本地和远端的重名变量做了去重操作（本地变量覆盖远端同名变量）    
4. 在vue文件中写样式的时候，通过`--`触发提示，如果没有提示，(1)重启编辑器试下，(2)检查下`.vscode/settings.json`配置是否不正确。    
5. 具体用法看demo图.    



CHANGELOG.md
<!--
 * @Author: wangyunbo
 * @Date: 2021-05-19 23:57:46
 * @LastEditors: wangyunbo
 * @LastEditTime: 2021-05-20 17:12:22
 * @Description: file content
 * @FilePath: \css-variables-hints\CHANGELOG.md
-->
# Change Log

All notable changes to the "css-variables-hints" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
/*
 * @Author: wangyunbo
 * @Date: 2021-05-19 23:57:46
 * @LastEditors: wangyunbo
 * @LastEditTime: 2021-05-21 17:03:30
 * @Description: file content
 * @FilePath: \css-variables-hints\src\extension.ts
 */
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import * as css from 'css';
import * as strip from 'strip-comments';
import { decode } from 'js-base64';

import type { Rule, Declaration } from 'css';
import axios from 'axios';

const { CompletionItem, CompletionList } = vscode

export async function activate(context: vscode.ExtensionContext) {
	// 存放提示(暂时未区分远端的资源和本地的资源，都放在一起)
	const items: vscode.CompletionItem[] = [];
	const bareItems: vscode.CompletionItem[] = [];
	const workspaceFolder = vscode.workspace.workspaceFolders || [];
	const folderPath = workspaceFolder[0]?.uri.fsPath;

	// workspace not selected
	if (!vscode.workspace.workspaceFolders) {
		return;
	}

	const config = vscode.workspace.getConfiguration('cssVarriablesHints');
	// 本地资源文件
	const hasFilesInConfig = config && config.has('files');
	// 远端资源文件
	const hasRemoteFilesInConfig = config && config.has('remoteFiles');

	// no config or specified files
	if (!config || !(hasFilesInConfig || hasRemoteFilesInConfig)) {
		return;
	}

	const configFiles: Array<string> = config.get('files') || [];
	const remoteConfigFiles: Array<string> = config.get('remoteFiles') || [];
	const configLanguageMods: Array<string> = config.get('languageModes') || [];
	// 处理本地变量
	handleConfigFiles(configFiles, folderPath, bareItems, items);
	// 处理远端变量文件
	const remoteFiles = await handleRemoteConfigFiles(remoteConfigFiles)
	remoteFiles.forEach(file => {
		handleCompletionItem(file, bareItems, items)
	})

	let completionProvider = vscode.languages.registerCompletionItemProvider(
		configLanguageMods.length ? configLanguageMods : ['css', 'postcss', 'scss', 'less', 'vue'],
		{
			async provideCompletionItems(document, position) {
				const firstCharOfLinePosition = new vscode.Position(position.line, 0);
				const withIndentBeforeCursorText = document.getText(new vscode.Range(firstCharOfLinePosition, position)) || '';
				const beforeCursorText = withIndentBeforeCursorText.trim();
				if (!beforeCursorText.match(/--([\w-]*)/)) {
					return null;
				}
				// 变量去重
				let dedup_bareItems = bareItems.reduce((finalItems: vscode.CompletionItem[], curItem: vscode.CompletionItem) => {
					if (!finalItems.some(item => item.label === curItem.label)) {
						finalItems.push(curItem)
					}
					return finalItems
				}, []);

				let dedup_items = items.reduce((finalItems: vscode.CompletionItem[], curItem: vscode.CompletionItem) => {
					if (!finalItems.some(item => item.label === curItem.label)) {
						finalItems.push(curItem)
					}
					return finalItems
				}, [])

				let completionItems: vscode.CompletionItem[] = beforeCursorText.match(/var\(--([\w-]*)/) ? dedup_bareItems : dedup_items;
				let dashPosition = null;
				let dashPosition2 = null;
				let range: vscode.Range | null = null;
				if (withIndentBeforeCursorText.endsWith('--')) {
					dashPosition = new vscode.Position(position.line, withIndentBeforeCursorText.lastIndexOf('--'));
					dashPosition2 = new vscode.Position(position.line, withIndentBeforeCursorText.lastIndexOf('--') + 2);
					range = new vscode.Range(dashPosition, dashPosition2);
				}
				completionItems = completionItems.map(item => {
					if (range === null || document.languageId !== 'vue') {
						return item
					} else {
						return {
							...item,
							additionalTextEdits: [vscode.TextEdit.delete(range)]
						}
					}
				})
				return new CompletionList(completionItems);
			},
		},
		'-', '-'
	);

	context.subscriptions.push(completionProvider);
}

function handleConfigFiles(configFiles: string[], folderPath: string, bareItems: vscode.CompletionItem[], items: vscode.CompletionItem[]) {
	configFiles.forEach((filePath) => {
		let file = fs.readFileSync(path.join(folderPath, filePath), { encoding: 'utf8' });
		handleCompletionItem(file, bareItems, items)
	});
}
// 处理远端文件（当前只针对gitlab）
async function handleRemoteConfigFiles(configFiles: any[]): Promise<any[]> {
	const rawRemoteFiles: any[] = [];
	for (const fileInfo of configFiles) {
		let { projectId = '', filePath = '', token = "", origin = "" } = fileInfo;
		filePath = encodeURIComponent(filePath)
		const url = `${origin}/api/v4/projects/${projectId}/repository/files/${filePath}?ref=master`;
		try {
			const result = await axios.get(url, {
				headers: {
					'PRIVATE-TOKEN': token
				}
			});
			if (result && result.data) {
				const { content: base64Content } = result.data;
				const fileContent = decode(base64Content);
				rawRemoteFiles.push(fileContent)
			}
		} catch (err) {
			continue;
		}
	};
	return rawRemoteFiles
}

// 对每一个资源文件生成CompletionItem
function handleCompletionItem(file: any, bareItems: vscode.CompletionItem[], items: vscode.CompletionItem[]) {
	// 去除资源里的注释
	file = strip(file);
	// 对资源文件做解析
	const cssParsed = css.parse(file);
	// 资源文件必须有根选择器
	const rootRule: Rule | undefined = cssParsed.stylesheet?.rules.find((rule: Rule) => {
		const isRuleType = rule.type = 'rule';
		const rootSelectors = [':export', ':root'];
		const hasRootSelector = rule?.selectors?.some(selector => rootSelectors.includes(selector));

		return Boolean(isRuleType && hasRootSelector);
	});
	const declarations = rootRule?.declarations;
	const variables = declarations?.filter((declaration: Declaration) => {
		return Boolean(
			declaration.type === 'declaration' &&
			declaration?.property?.startsWith('--')
		);
	});

	variables?.forEach((variable: Declaration) => {
		// For use when the user has already typed `var(`
		const completionItemBare = new CompletionItem(variable.property!, vscode.CompletionItemKind.Variable);
		// 变量的值
		completionItemBare.detail = variable.value;
		// 变量的属性
		completionItemBare.insertText = variable.property;
		bareItems.push(completionItemBare);

		const completionItem = new CompletionItem(variable.property!, vscode.CompletionItemKind.Variable);

		completionItem.detail = variable.value;
		completionItem.insertText = `var(${variable.property})`;
		completionItem.keepWhitespace = true;

		items.push(completionItem);
	});
}